drivers: gpio: Add APIs for enabling/disabling interrupt

LSChyi opened the pull request:
Link to issue #52186

This PR is for the proposed new APIs for enabling/disabling interrupt in GPIO drivers, and also builds an implementation for it8xxx2 as an example.

New APIs for enabling/disabling interrupt in GPIO drivers is backward compatible, which means the original API `pin_interrupt_configure` to modify the  interrupt settings is still valid, and can turn on/off the interrupt as well. The proposed new APIs `pin_interrupt_enable` and `pin_interrupt_disable` are not used for configuring the interrupt settings, and is used only for turning on/off the interrupt. The interrupt configuration is still done by the original API `pin_interrupt_configure`.

LSChyi committed:
drivers: gpio: Add APIs for enabling/disabling interrupt

Add pin_interrupt_enable and pin_interrupt_disable in gpio_driver_api,
and add corresponding APIs in gpio.h for application to enable/disable
an interrupt without reconfiguring again.
This CL also Create a new Kconfig option for this feature.

Signed-off-by: Sung-Chi Li <lschyi@google.com>

LSChyi committed:
drivers: gpio: it8xxx2: Implement enable/disable pin interrupt

Implement the GPIO_INT_MODE_ENABLE_ONLY and GPIO_INT_MODE_DISABLE_ONLY
on it8xxx2 gpio driver.

Signed-off-by: Sung-Chi Li <lschyi@google.com>

cfriedt commented on include/zephyr/drivers/gpio.h:
```diff
@@ -538,9 +538,14 @@ __subsystem struct gpio_driver_api {
 				       gpio_pin_t pin,
 				       enum gpio_int_mode, enum gpio_int_trig);
+        /* Enable the pin interrupt. */
+	int (*pin_interrupt_enable)(const struct device *port, gpio_pin_t pin);
```
There needs to be a plan to implement this for the other ~2 dozen GPIO drivers otherwise the implementations inconsistent / fragmented.

Also, tests need to be added, etc.

LSChyi replied:
Thank you! I put it as a experimental Kconfig, and add tests for it.

keith-zephyr replied:
Please see my [comment](https://github.com/zephyrproject-rtos/zephyr/issues/52186#issuecomment-1332576434) on the RFC.  I think you can avoid expanding the driver side API.

LSChyi replied:
Followed comments in the RFC, I think current approach does not changing the driver API. However, I found that maybe I can remove this experimental KConfig, and maybe just add guarding check in the ~2 dozen GPIO drivers, and return `-ENOTSUP` for drivers not supported yet?

cfriedt reviewed (changes requested):
This will introduce driver inconsistencies unless there are additional implementations made.

Also, there are no tests.

It would be smart to introduce this with a Kconfig / EXPERIMENTAL.

LSChyi committed:
tests: drivers: gpio: Add test for GPIO_ENABLE_DISABLE_INTERRUPT

Add test for the experimental feature GPIO_ENABLE_DISABLE_INTERRUPT,
which covers APIs gpio_pin_interrupt_enable() and
gpio_pin_interrupt_disable().

Signed-off-by: Sung-Chi Li <lschyi@google.com>

LSChyi committed:
drivers: gpio: Implement pin interrupt enable and disable

Implement the driver method pin_interrupt_enable and pin_interrupt_disable.
This commit fixes getting the get_pending_int by updating the interrupts
field in the gpio_emul_data filed when interrupt is triggered. Also,
introduces a new filed enabled_interrupts to better simulate the
behavior of the interrupt pending and whether the interrupt is
enabled/disabled.

Signed-off-by: Sung-Chi Li <lschyi@google.com>

henrikbrixandersen commented:
> @henrikbrixandersen - what are your thoughts here?

Please see my comment on the issue: https://github.com/zephyrproject-rtos/zephyr/issues/52186#issuecomment-1334410246

carlescufi commented:
@cfriedt ping

cfriedt commented:
> > @henrikbrixandersen - what are your thoughts here?
> 
> Please see my comment on the issue: https://github.com/zephyrproject-rtos/zephyr/issues/52186#issuecomment-1334410246

It seems that the path of least resistance would be to add some doxygen to specify the desired behaviour of configure_interrupt.

I think it would likely require adding something to the testsuite.

The difficulty is that we don't have a platform-independent `is_gpio_interrupt_pending()` function, but it would be possible to add that functionality in the testsuite as a test-specific syscall. Yet more fun is that an implementation would need to be added per platform, unless I'm mistaken.

It would likely be better to make increase the specificity of the API rather than add another function call to asynchronously enforce desired behaviour in a handful of products.

keith-zephyr commented on tests/drivers/gpio/gpio_enable_disable_interrupt/dts/bindings/test-gpio-enable-disable-interrupt.yaml:
```diff
@@ -0,0 +1,18 @@
+    This binding provides resources required to build and run the
+    tests/drivers/gpio/gpio_enable_disable_interrupt test in Zephyr.
+
+compatible: "test-gpio-enable-disable-interrupt"
```
I see two patterns for defining devicetree nodes used by tests.  Some tests define the schema using the `vnd,` prefix under [dts/bindings/tests](https://github.com/zephyrproject-rtos/zephyr/tree/main/dts/bindings/test) and a few tests create local dts/bindings folder like you did here.

It looks like it's more common use `vnd,` prefix.  Can you move this?

LSChyi replied:
Done

keith-zephyr commented on tests/drivers/gpio/gpio_enable_disable_interrupt/src/main.c:
```diff
@@ -0,0 +1,100 @@
+		   "failed to set interrupt with edge rising");
+	disable_interrupt(fixture->port, fixture->pin);
+	trigger_callback(fixture->port, fixture->pin);
+	zassert_false(cb_called, "callback should not executed before configuring the interrupt");
```
```suggestion
	zassert_false(cb_called, "callback should not executed before enable the interrupt");
```

LSChyi replied:
Done

keith-zephyr commented on include/zephyr/drivers/gpio.h:
```diff
@@ -129,6 +129,14 @@ extern "C" {
+ * This is a component flag that should be combined with `GPIO_INT_ENABLE` or
+ * `GPIO_INT_DISABLE` flags to produce a meaningful configuration.
+ */
+#define GPIO_INT_ENABLE_DISABLE_ONLY   (1u << 27)
```
This should be guarded with CONFIG_GPIO_ENABLE_DISABLE_INTERRUPT.  So that applications will get a compile error if trying to use this bit without the config enabled.

LSChyi replied:
Done

keith-zephyr commented:
> > > @henrikbrixandersen - what are your thoughts here?
> > 
> > 
> > Please see my comment on the issue: [#52186 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/52186#issuecomment-1334410246)
> 
> It seems that the path of least resistance would be to add some doxygen to specify the desired behaviour of configure_interrupt.
> 
> I think it would likely require adding something to the testsuite.
> 
> The difficulty is that we don't have a platform-independent `is_gpio_interrupt_pending()` function, but it would be possible to add that functionality in the testsuite as a test-specific syscall. Yet more fun is that an implementation would need to be added per platform, unless I'm mistaken.
> 
> It would likely be better to make increase the specificity of the API rather than add another function call to asynchronously enforce desired behaviour in a handful of products.

The test provided in the latest PR does a good job of validating the new behavior with the generic GPIO emulator.  However, the new flag `GPIO_INT_ENABLE_DISABLE_ONLY` isn't possible to test with other GPIO driver instances to verify consistent behavior.  That would require a driver specific emulator written for every GPIO driver.

cfriedt commented:
> The test provided in the latest PR does a good job of validating the new behavior with the generic GPIO emulator. However, the new flag `GPIO_INT_ENABLE_DISABLE_ONLY` isn't possible to test with other GPIO driver instances to verify consistent behavior. That would require a driver specific emulator written for every GPIO driver.

An emulator for every GPIO driver is obviously overkill. I was suggesting that a test case should be added that verifies specified behaviour when an interrupt is already pended. For that, implementation specific details could be added to the test incrementally.

@henrikbrixandersen - please chime in if there is anything specific you would like done with this PR.

keith-zephyr commented on include/zephyr/drivers/gpio.h:
```diff
@@ -129,6 +129,16 @@ extern "C" {
+ * This is a component flag that should be combined with `GPIO_INT_ENABLE` or
+ * `GPIO_INT_DISABLE` flags to produce a meaningful configuration.
+ */
+#define GPIO_INT_ENABLE_DISABLE_ONLY   (1u << 27)
```
Please also update the documentation for GPIO [interrupt configuration flags](https://github.com/zephyrproject-rtos/zephyr/blob/e2742af6ed1118e1e2dd2b0ffe89ce5c210ab22d/include/zephyr/drivers/gpio.h#L82)  to provide details about the behavior of pending interrupts.

Something like:
"The GPIO controller should clear any pending interrupt on the pin when changing the triggering properties.  Applications should use the GPIO_INT_MODE_DISABLE_ONLY and GPIO_INT_MODE_ENABLE_ONLY flags to enable and disable interrupts on the pin without changing any GPIO settings."

galak commented:
Dev-review (Dec 15th, 2022): Didn't feel like we had the correct people in the meeting to close on this and suggested to raise in Arch Review.

Kumar brought up the general pattern of how do we handle gray areas of APIs (should also discuss in the Arch Meeting).

carlescufi commented:
Architecture WG:

- @keith-zephyr mentions that they have a need to enable and disable the GPIO interrupt, without changing anything else
- The behavior when an interrupt is disabled while a GPIO is pending is not currently defined, so this needs to be clarified
- @cfriedt supports the idea of this functionality, he would like to see it working for more than one platform before merging
- Greg Shue asks whether this could be generalized but Brix explains that it's not possible due to the nature of GPIOs
- @henrikbrixandersen aligns with a positive opinion of the change

Conclusion: An implementation will be added for more SoC families, and no further objections.

keith-zephyr commented:
> Architecture WG:
> 
> * @keith-zephyr mentions that they have a need to enable and disable the GPIO interrupt, without changing anything else
> * The behavior when an interrupt is disabled while a GPIO is pending is not currently defined, so this needs to be clarified
> * @cfriedt supports the idea of this functionality, he would like to see it working for more than one platform before merging
> * Greg Shue asks whether this could be generalized but Brix explains that it's not possible due to the nature of GPIOs
> * @henrikbrixandersen aligns with a positive opinion of the change
> 
> Conclusion: An implementation will be added for more SoC families, and no further objections.

From the architecture meeting: implementing STM32, Nuvton, ITE, and the emulator would be sufficient for this change.

besmarsh commented:
Not sure if this has already been considered here, but another side-effect of using `pin_interrupt_configure` to enable/disable interrupts is that configurations are lost after configuring, disabling, and then re-enabling. For example:
```
gpio_pin_interrupt_configure(port, pin, GPIO_INT_EDGE_RISING));
gpio_pin_interrupt_configure(port, pin, GPIO_INT_DISABLE);
gpio_pin_interrupt_configure(port, pin, GPIO_INT_ENABLE);
```

Running this will cause the rising edge configuration to be lost after disabling and then re-enabling. I expect this is the intended behaviour. However, the new `GPIO_INT_ENABLE_DISABLE_ONLY` should ensure that previous configuration such as this is retained.

LSChyi committed:
drivers: gpio: stm32: Implement enable/disable pin interrupt

Implement the GPIO_INT_MODE_ENABLE_ONLY and GPIO_INT_MODE_DISABLE_ONLY
on stm32 GPIO driver.

Signed-off-by: Sung-Chi Li <lschyi@google.com>

LSChyi committed:
drivers: gpio: numicro: Implement enable/disable pin interrupt

Implement the GPIO_INT_MODE_ENABLE_ONLY and GPIO_INT_MODE_DISABLE_ONLY
on numicro GPIO driver.

Signed-off-by: Sung-Chi Li <lschyi@google.com>

LSChyi commented:
Hi @besmarsh, I found it is not that useful to implement such API on all STM32 series, for we have no chance to keep the pending register react to the GPIO input change, yet not trigger the interrupt (refer to [RM0090](https://www.st.com/resource/en/reference_manual/DM00031020-.pdf) for STM32F4 series).
Do we have some STM32 chips to target on?

besmarsh commented:
> Hi @besmarsh, I found it is not that useful to implement such API on all STM32 series, for we have no chance to keep the pending register react to the GPIO input change, yet not trigger the interrupt (refer to [RM0090](https://www.st.com/resource/en/reference_manual/DM00031020-.pdf) for STM32F4 series). Do we have some STM32 chips to target on?

There is no need to have the pending register react to the GPIO input change but not trigger the interrupt (when disable only is used). I'd just like to see the trigger configuration retained when disable only is used and followed by enable only.

> For the existing [STM32 GPIO driver](https://github.com/zephyrproject-rtos/zephyr/blob/c2e4029aad0a45b36449fd95a45e79e8280d4556/drivers/gpio/gpio_stm32.c#L594), can't you just make calls to `stm32_exti_disable()` and `stm32_exti_enable()` when the `GPIO_INT_ENABLE_DISABLE_ONLY` flag is set?

I think something along these lines would be sufficient. Whereas the current STM32 GPIO driver sets the trigger (`stm32_exti_trigger()`), for disable only and enable only this could be skipped and the driver could only call `stm32_exti_disable()` and `stm32_exti_enable()`.

@LSChyi I could contribute the changes to the STM32 GPIO driver to support this if you wish, I don't want to intrude on your feature though if you want to do it :)

LSChyi committed:
driver: gpio: npcx: Implement enable/disable pin interrupt

Implement the GPIO_INT_MODE_ENABLE_ONLY and GPIO_INT_MODE_DISABLE_ONLY
on npcx GPIO driver.

Signed-off-by: Sung-Chi Li <lschyi@google.com>

LSChyi commented:
Thank you @keith-zephyr and @besmarsh, then it should not be a problem in stm32, I can do it.

keith-zephyr commented on drivers/gpio/gpio_emul.c:
```diff
@@ -652,15 +665,31 @@ static int gpio_emul_pin_interrupt_configure(const struct device *port, gpio_pin
 		drv_data->flags[pin] &= ~GPIO_EMUL_INT_BITMASK;
 		drv_data->flags[pin] |= GPIO_INT_DISABLE;
+#ifdef CONFIG_GPIO_ENABLE_DISABLE_INTERRUPT
+	case GPIO_INT_MODE_DISABLE_ONLY:
```
You need to document fallthrough conditions.
```suggestion
	__fallthrough;
	case GPIO_INT_MODE_DISABLE_ONLY:
```

LSChyi replied:
Done

keith-zephyr commented on drivers/gpio/gpio_emul.c:
```diff
@@ -652,15 +665,31 @@ static int gpio_emul_pin_interrupt_configure(const struct device *port, gpio_pin
 		drv_data->flags[pin] &= ~GPIO_EMUL_INT_BITMASK;
 		drv_data->flags[pin] |= (mode | trig);
+#ifdef CONFIG_GPIO_ENABLE_DISABLE_INTERRUPT
+	case GPIO_INT_MODE_ENABLE_ONLY:
```
```suggestion
	__fallthrough;
	case GPIO_INT_MODE_ENABLE_ONLY:
```

LSChyi replied:
Done

keith-zephyr commented on tests/drivers/gpio/gpio_enable_disable_interrupt/boards/native_posix.overlay:
```diff
@@ -0,0 +1,16 @@
+/ {
+	resources {
+		compatible = "test-gpio-enable-disable-interrupt";
+		in-gpios = <&gpio0 0 0>; /* Pin 0 */
```
Nit - name `irq-gpios` to follow general Zephyr convention.

LSChyi replied:
Done

keith-zephyr commented on tests/drivers/gpio/gpio_enable_disable_interrupt/src/main.c:
```diff
@@ -0,0 +1,100 @@
+}
+
+struct gpio_enable_disable_interrupt_fixture {
+	const struct device *port;
+	gpio_pin_t pin;
```
You can use `struct gpio_dt_spec` here instead and then call `gpio_pin_configure_dt()` and `gpio_pin_interrupt_configure_dt()`.  

@cfriedt is it preferred for applications/samples to use `gpio_dt_spec` in place of separate `struct device` and `gpio_pin_t` fields?

LSChyi replied:
I updated this in another commit, if @cfriedt is also ok with that, then I will squash that commit.

keith-zephyr replied:
@cfriedt will be unavailable for a week or so.  But I think using `gpio_dt_spec` results in a cleaner implementation, so I think you should squash the changes.

keith-zephyr commented on tests/drivers/gpio/gpio_enable_disable_interrupt/src/main.c:
```diff
@@ -0,0 +1,100 @@
+	struct gpio_enable_disable_interrupt_fixture *fixture =
+		(struct gpio_enable_disable_interrupt_fixture *)arg;
+
+	zassert_true(device_is_ready(fixture->port), "GPIO device is not ready");
```
If you switch to using `gpio_dt_spec`, this call becomes `gpio_is_ready_dt()`

keith-zephyr commented on tests/drivers/gpio/gpio_enable_disable_interrupt/src/main.c:
```diff
@@ -0,0 +1,100 @@
+
+	zassert_ok(gpio_pin_interrupt_configure(fixture->port, fixture->pin, GPIO_INT_DISABLE));
+	gpio_init_callback(&cb_data, callback, BIT(fixture->pin));
+	zassert_ok(gpio_add_callback(fixture->port, &cb_data), "failed to add callback");
```
You should create an "after" function that calls `gpio_remove_callback()`.

LSChyi replied:
Thank you for catching this issue, fixed.

keith-zephyr reviewed (changes requested):
Note - I reviewed everything by the numicro driver change.

LSChyi commented on drivers/gpio/gpio_emul.c:
```diff
@@ -442,6 +450,7 @@ static int gpio_emul_pin_configure(const struct device *port, gpio_pin_t pin,
 
 	k_mutex_unlock(&drv_data->mu);
 	gpio_fire_callbacks(&drv_data->callbacks, port, BIT(pin));
+	drv_data->interrupts &= ~((gpio_port_pins_t)BIT(pin));
```
Done

LSChyi commented on drivers/gpio/gpio_emul.c:
```diff
@@ -652,15 +665,31 @@ static int gpio_emul_pin_interrupt_configure(const struct device *port, gpio_pin
 
+#ifdef CONFIG_GPIO_ENABLE_DISABLE_INTERRUPT
+	if (!(mode & GPIO_INT_ENABLE_DISABLE_ONLY)) {
+		drv_data->interrupts &= ~((gpio_port_pins_t)BIT(pin));
```
Done

LSChyi commented on drivers/gpio/gpio_emul.c:
```diff
@@ -627,7 +636,11 @@ static int gpio_emul_pin_interrupt_configure(const struct device *port, gpio_pin
 		return -EINVAL;
 	}
 
+#ifdef CONFIG_GPIO_ENABLE_DISABLE_INTERRUPT
+	if (mode != GPIO_INT_MODE_DISABLED && !(mode & GPIO_INT_ENABLE_DISABLE_ONLY)) {
+#else
 	if (mode != GPIO_INT_MODE_DISABLED) {
+#endif /* CONFIG_GPIO_ENABLE_DISABLE_INTERRUPT */
```
We have a case when calling the `gpio_pin_interrupt_configure` with flags is `GPIO_INT_MODE_ENABLE_ONLY`, the `trig` is not specified as `GPIO_INT_TRIG_LOW`, `GPIO_INT_TRIG_HIGH`, or `GPIO_INT_TRIG_BOTH`. In this case, if we cannot and does not need to check the input configuration parameter, thus we need to check the `GPIO_INT_ENABLE_DISABLE_ONLY ` bit in this case.
I also tried remove this check and run the test, the test also failed.

keith-zephyr reviewed (commented):
Looks good, but this needs a rebase.

keith-zephyr commented:
@ssekar15 Can you review the numicro driver change?

keith-zephyr commented:
@MulinChao and @ChiHuaL  can you review the NPCX driver change?

keith-zephyr commented:
@Dino-Li can you review the ITE driver change?

MulinChao commented:
> @MulinChao and @ChiHuaL can you review the NPCX driver change?

Hi @LSChyi and @keith-zephyr:
Regarding the changes in NPCX gpio driver, it looks good to me. Thanks for the effort.

Dino-Li commented:
> @Dino-Li can you review the ITE driver change?

LGTM. +1 for it8xxx2 gpio driver changes.

cfriedt commented:
Thanks for working on this. I'll give it a quick once-over when CI is finished.

cfriedt commented on tests/drivers/gpio/gpio_enable_disable_interrupt/testcase.yaml:
```diff
@@ -0,0 +1,5 @@
+  peripheral.gpio.enable_disable_interrupt:
+    tags: drivers gpio
+    depends_on: gpio
+    filter:  dt_compat_enabled("test-gpio-enable-disable-interrupt")
```
Great use of the `filter` property 👍

cfriedt commented on drivers/gpio/gpio_emul.c:
```diff
@@ -92,6 +92,10 @@ struct gpio_emul_data {
+	/** Is interrupt enabled for each pin */
+	gpio_port_pins_t enabled_interrupts;
+	/** Mutex to synchronize accesses to driver data and config */
+	struct k_mutex mu;
```
It looks like there is a conflict here. The old locking mechanism (mutex mu), was replaced with a new one (spinlock lock) somewhat recently. Please update

LSChyi replied:
Thank you for catching this. Removed unused `k_mutex`, and I think changes in this file should work with the spin lock.

keith-zephyr commented on tests/drivers/gpio/gpio_enable_disable_interrupt/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(gpio_basic_api)
+
+FILE(GLOB app_sources src/*.c)
```
Note - [cmake documentation](https://cmake.org/cmake/help/latest/command/file.html?highlight=CONFIGURE_DEPENDS#filesystem) discourages using GLOB.

LSChyi replied:
Updated to only add the target `src/main.c` instead of using `GLOB`.

keith-zephyr commented on drivers/gpio/gpio_numicro.c:
```diff
@@ -162,10 +166,23 @@ static int gpio_numicro_pin_interrupt_configure(const struct device *dev,
+#ifdef CONFIG_GPIO_ENABLE_DISABLE_INTERRUPT
+	if (mode == GPIO_INT_MODE_DISABLE_ONLY) {
+		cfg->regs->INTEN &= ~(BIT(pin) | BIT(pin + 16));
+		return 0;
```
Note - you're not updating the `data->interrupt_en_reg` cached copy in this (or the INT_MODE_ENABLE_ONLY) path.

LSChyi replied:
This is the enable and disable only path, so in enabling branch, we restore the `INTEN` register by referencing what we recorded in `interrupt_en_reg`; in disabling branch, we simply disable the `INTEN` register. Thus, there is no need to update the `interrupt_en_reg` here. Maybe I should use a better naming for this register.

LSChyi commented on drivers/gpio/gpio_numicro.c:
```diff
@@ -36,6 +36,10 @@ struct gpio_numicro_data {
 	sys_slist_t callbacks;
+#ifdef CONFIG_GPIO_ENABLE_DISABLE_INTERRUPT
+	/* backup of the INTEN register */
+	uint32_t interrupt_en_reg;
```
Numicro uses the M478JIDAE, so from the section 6.6.7 [datasheet](https://www.nuvoton.com/export/resource-files/en-us--TRM_M480_Series_EN_Rev3.03.pdf), the `INTEN` register is split into higher half and lower half for enabling/disabling for the rising and falling edge trigger respectively, so we need just an extra `uint32t` for recording the enabled information per port.
I can add some comments here to better illustrate the purpose of this register.

cfriedt commented:
Removed the `DNM` label - thanks for the additional driver implementations 👍

LSChyi commented:
Thank you all!
